src/bsp/src/stm32/src/Decawave.cpp

```diff
@@ -0,0 +1,307 @@
+#include <DecawaveUtils.h>
+#include <FreeRTOS.h>
+#include <Task.h>
+#include <cstring>
+#include <deca_device_api.h>
+#include <deca_regs.h>
+#include <task.h>
+
+void rxCallback(const dwt_cb_data_t* callbackData, void* context) {
+    memcpy(&(static_cast<Decawave*>(context)->m_callbackData), callbackData, sizeof(dwt_cb_data_t));
+
+    BaseType_t taskWoken = pdFALSE;
+
+    if (static_cast<Decawave*>(context)->m_rxTaskHandle != NULL) {
+        vTaskNotifyGiveFromISR(static_cast<Decawave*>(context)->m_rxTaskHandle, &taskWoken);
+    }
+
+    static_cast<Decawave*>(context)->m_rxTaskHandle = NULL;
+    portYIELD_FROM_ISR(taskWoken);
+}
+
+void isrCallback(void* context) {
+    deca_selectDevice(static_cast<Decawave*>(context)->m_spiDevice);
+    dwt_isr();
+}
+
+void rxAsyncTask(void* context) {
+    while (true) {
+        ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+        static_cast<Decawave*>(context)->retrieveRxFrame(
+            static_cast<Decawave*>(context)->m_rxFrame);
+    }
+}
+
+Decawave::Decawave(decaDevice_t spiDevice) :
+    m_spiDevice(spiDevice),
+    m_channelNo(2),
+    m_speed(UWBSpeed::SPEED_110K),
+    m_rxAsyncTask("dw_rx_task", tskIDLE_PRIORITY + 10, rxAsyncTask, this) {}
+
+Decawave::Decawave(decaDevice_t spiDevice, uint8_t channel, UWBSpeed speed) :
+    m_spiDevice(spiDevice),
+    m_speed(speed),
+    m_rxAsyncTask("dw_rx_task", tskIDLE_PRIORITY + 10, rxAsyncTask, this) {
+
+    if (channel > 0 && channel < 8) {
+        m_channelNo = channel;
+    }
+}
+
+bool Decawave::start() {
+    deca_selectDevice(m_spiDevice);
+    deca_setSlowRate();
+    uint32_t deviceID = 0;
+    uint8_t i = 0;
+    // Retry to read the deviceID 10 times before abandoning
+    while (deviceID != DWT_DEVICE_ID && i++ < 10) {
+        deviceID = dwt_readdevid();
+        Task::delay(1);
+    }
+
+    if (dwt_initialise(DWT_LOADUCODE) == DWT_ERROR) {
+        return false;
+    }
+
+    deca_selectDevice(m_spiDevice);
+    deca_setISRCallback(m_spiDevice, isrCallback, this);
+    dwt_setcallbacks(rxCallback, rxCallback, rxCallback, rxCallback, this);
+
+    deca_setFastRate();
+
+    dwt_softreset();
+    configureDW();
+
+    dwt_enablegpioclocks();
+    dwt_setgpiodirection(DWT_GxM0 | DWT_GxM1 | DWT_GxM2 | DWT_GxM3, 0);
+
+    dwt_setinterrupt(DWT_INT_RFCG | DWT_INT_RFTO | DWT_INT_RXPTO | DWT_INT_RPHE | DWT_INT_RFCE |
+                         DWT_INT_RFSL | DWT_INT_SFDT,
+                     1);
+
+    setLed(DW_LED::LED_0, true);
+
+    m_rxAsyncTask.start();
+
+    return true;
+}
+
+bool Decawave::setChannel(uint8_t channelNo) {
+    if (channelNo < 1 || channelNo > 7) {
+        return false;
+    }
+
+    m_channelNo = channelNo;
+    configureDW();
+    return true;
+}
+
+void Decawave::setSpeed(UWBSpeed speed) {
+    m_speed = speed;
+    configureDW();
+}
+
+void Decawave::setLed(DW_LED led, bool enabled) {
+    uint8_t dwGPIO = 0;
+    uint8_t dwGPIOValue = 0;
+
+    switch (led) {
+    case DW_LED::LED_0:
```

Comment:
would'nt be simpler to have LED_x = DWT_GxPx and remove the switch/case?

Reply:
It wasn't simpler because C++ class enums can't directly convert to their int value, but I added a utils function to do it. It's much cleaner. Also removed the conversion between our speed enum and the DW macros